Allow visibility on constructors

AbubakerB opened the pull request:
Fixes #2341.

Changes:
1 - Allows private, protected and public accessibilities on constructors.
2 - Disallows classes with private constructors from being extended.
3 - Now throws an error on constructor overloads with different access modifiers (than the constructors implementation).

msftclas commented:
Hi __@AbubakerB__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, MSBOT;

RyanCavanaugh commented:
For this to work properly, the visibility needs to be part of the construct signature rather than being a declaration-based check that happens during signature resolution. Otherwise it's going to be too difficult to enforce assignability constraints based on constructor visibility.

A relevant testcase to add would be:
```ts
class Foo {
  private constructor() { }
}
class Bar {
  protected constructor() { }
}
let x: new() => any;
x = Foo; // Should be an error
x = Bar; // Should be an error
```

AbubakerB commented:
@RyanCavanaugh SignaturesRelatedTo now checks visibility of constructor.

AbubakerB commented:
Should i close this off? or is this in the pipeline?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -8697,6 +8723,37 @@ namespace ts {
+            
+            function checkConstructorVisibility(signature: Signature, node: NewExpression) {
+                let constructor = result.declaration;                
+                if (!constructor || !node.expression) return;
```
Can you wrap these `return` statements with curlies and place them on their own line?

```TypeScript
if (!constructor || !node.expression) {
    return;
}
```

AbubakerB replied:
Done

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -8697,6 +8723,37 @@ namespace ts {
+            function checkConstructorVisibility(signature: Signature, node: NewExpression) {
+                let constructor = result.declaration;                
+                if (!constructor || !node.expression) return;
+                // if constructor is public, we dont need to check visibility
```
dont -> don't

AbubakerB replied:
Done

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -5097,6 +5097,13 @@ namespace ts {
                 if (sourceSig && targetSig) {
+                    if (kind === SignatureKind.Construct) { 
+                        result &= signatureVisibilityRelatedTo(sourceSig, targetSig);
+                        if (result !== Ternary.True) {
```
Maybe I'm missing something shouldn't this be `===`?

AbubakerB replied:
I thought I should follow the same structure as [here](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts#L5087).

It checks to see if its not assignable (resulted from ```signatureVisibilityRelatedTo```) and returns if not.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -15003,7 +15054,25 @@ namespace ts {
+                                    let signatures = getSignaturesOfType(baseType, SignatureKind.Construct);
+                                    for (let signature of signatures) {
+                                        let constuctor = signature.declaration;
+                                        if (constuctor && constuctor.flags & NodeFlags.Private) {
```
Wrap `constuctor.flags & NodeFlags.Private` with parentheses

AbubakerB replied:
Done

DanielRosenwasser commented:
The declaration emitter will need to be amended - add `// @declaration: true` to the top of every file.

DanielRosenwasser commented:
The declaration emitter will need to be amended - add `// @declaration: true` to the top of every affected test file.

DanielRosenwasser commented:
The following test cases would be useful too:

```TypeScript
class Base {
    protected constructor() {
    }
}

class Derived extends Base {
    private constructor() {
        super();
    }
}
```

```TypeScript
class Base {
    protected constructor() {
    }
}

class Derived extends Base {
    protected constructor() {
        super();
    }
}

var baseCtor = Base;
baseCtor = Derived;
```

DanielRosenwasser commented:
@AbubakerB sorry this fell off of our radar, we've been hard at work on TypeScript 1.6.

@ahejlsberg @mhegazy @vladima any thoughts?

AbubakerB commented:
@DanielRosenwasser I see, that's fine :)

1. I've added ```// @declaration: true``` to the three test files.
2. Added test cases under ```classConstructorAccessibility2.ts```. Although, I have one question... Can a X ctor signature be assigned to a non-X ctor signature? (where X is public/protected/private).
3. Renamed the diagnostic message for error code 2656 to be more clear.

DanielRosenwasser commented on tests/baselines/reference/classConstructorAccessibility.errors.txt:
```diff
@@ -1,29 +1,39 @@
-tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(10,5): error TS1089: 'protected' modifier cannot appear on a constructor declaration.
-tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(23,9): error TS1089: 'private' modifier cannot appear on a constructor declaration.
-tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(27,9): error TS1089: 'protected' modifier cannot appear on a constructor declaration.
+tests/cases/conformance/classes/constructorDeclarations/classConstructorAccessibility.ts(20,9): error TS2655: Constructor '(x: number): D' is private and only accessible within class 'D'.
```
Shouldn't this be `new (x: number): D`?

AbubakerB replied:
What would be the best way of adding the ```new``` before the ```signature``` output?
I can't seem to find any method that writes ```'new' + signature```.

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1697,6 +1697,22 @@
+        "category": "Error",
+        "code": 2654
+    },
+    "Constructor '{0}' is private and only accessible within class '{1}'.": {
```
Maybe it should be "of type `'{0}'`

AbubakerB replied:
Done

RyanCavanaugh commented:
@AbubakerB can you get this up-to-date and resubmit in a new PR? Very sorry we left this hanging for so long.

AbubakerB commented:
Sure. I've opened a new PR for this.
